src/misc/unicode.cpp

```diff
@@ -1284,6 +1679,137 @@ static void construct_decomposed(const code_page_mapping_t& normalized,
+
+		bool is_group_profitable = false;
+		for (const auto& [from, target] : out_box_code_points) {
+			for (const auto& alias : alias_group) {
+				if (from != alias.first) {
+					continue;
+				}
+
+				if (!code_page_mapping.count(alias.second)) {
+					// We cannot apply this alias group
+					return false;
+				}
+
+				if (!code_page_mapping.count(target)) {
+					is_group_profitable = true;
+				}
+			}
+		}
+
+		return is_group_profitable;
+	};
+
+	auto try_set = [&](const std::vector<uint16_t>& drawing_set) {
+		assert(box_drawing_set_regular.size() == drawing_set.size());
+
+		// Create initial mapping
+		out_box_code_points.clear();
+		for (size_t idx = 0; idx < drawing_set.size(); ++idx) {
+			const auto code_point_1 = box_drawing_set_regular[idx];
+			const auto code_point_2 = drawing_set[idx];
+			out_box_code_points[code_point_1] = code_point_2;
+		}
+
+		// If necessary, adapt mapping using alias groups
+		for (const auto& alias_group : box_alias_groups) {
+			if (!is_adaptation_needed()) {
+				// No adaptation needed - our job is done here
+				return true;
+			}
+
+			if (!is_alias_group_suitable(alias_group)) {
+				continue;
+			}
+
+			// Apply the aliases from the current group
+			for (const auto& [from, target] : out_box_code_points) {
+				for (const auto& alias : alias_group) {
+					if (from != alias.first) {
+						continue;
+					}
+
+					out_box_code_points[from] = alias.second;
+				}
+			}
+		}
+
+		return !is_adaptation_needed();
+	};
+
+	// Try to adjust regular (full) drawing character set for the code page,
+	// if failed try the reduced set without double lines
+	if (try_set(box_drawing_set_regular) || try_set(box_drawing_set_light)) {
+		return;
+	}
+
+	// Last resort fallback - use 7-bit ASCII fallback for everything
+	out_box_code_points.clear();
+	for (const auto& code_point : box_drawing_set_regular) {
+		if (mapping_ascii.count(code_point)) {
+			out_box_code_points[code_point] = mapping_ascii.at(code_point);
+		} else {
+			out_box_code_points[code_point] = unknown_character;
+		}
+	}
+}
+
+static void construct_case_mapping(const map_code_point_case_t& map_case_global,
+		                   const map_grapheme_to_dos_t& code_page_mapping,
+		                   map_dos_character_case_t& out_map_case)
+{
+	out_map_case.resize(UINT8_MAX);
+	for (uint16_t idx = 0; idx < UINT8_MAX; ++idx) {
+		if (idx < decode_threshold_non_ascii && map_case_global.count(idx)) {
+			out_map_case[idx] = static_cast<uint8_t>(
+				map_case_global.at(idx));
+		} else {
+			out_map_case[idx] = static_cast<uint8_t>(idx);
+		}
+	}
+
+	for (const auto& [grapheme, character_code] : code_page_mapping) {
+		if (!map_case_global.count(grapheme.GetCodePoint())) {
+			// No information how to switch case for this code point
+			continue;
+		}
+
+		Grapheme grapheme_switched = map_case_global.at(
+			grapheme.GetCodePoint());
+		grapheme_switched.CopyMarksFrom(grapheme);
+
+		if (!code_page_mapping.count(grapheme_switched)) {
```

Comment:
Suggest quantity comparison (count == 0) otherwise as a bool, it looks more like a success/fail check, such as "did we have an error counting elements?"

With C++20, we'll finally get `.contains(key)`

Otherwise as a bool

Reply:
OK, changed